clients/drcachesim/scheduler/scheduler_impl.cpp

```diff
@@ -1385,63 +1395,98 @@ template <typename RecordType, typename ReaderType>
+{
+    if (!trace_path.empty()) {
+        reader = get_reader(trace_path, verbosity_);
         if (!reader || !reader->init()) {
-            error_string_ +=
-                "Failed to open kernel switch file " + options_.kernel_switch_trace_path;
+            error_string_ += "Failed to open file for kernel " + sequence_type +
+                " sequences: " + trace_path;
             return sched_type_t::STATUS_ERROR_FILE_OPEN_FAILED;
         }
         reader_end = get_default_reader();
-    } else if (!options_.kernel_switch_reader) {
+    } else if (!reader) {
         // No switch data provided.
         return sched_type_t::STATUS_SUCCESS;
     } else {
-        if (!options_.kernel_switch_reader_end) {
-            error_string_ += "Provided kernel switch reader but no end";
+        if (!reader_end) {
+            error_string_ += "Provided kernel " + sequence_type + " reader but no end";
             return sched_type_t::STATUS_ERROR_INVALID_PARAMETER;
         }
-        reader = std::move(options_.kernel_switch_reader);
-        reader_end = std::move(options_.kernel_switch_reader_end);
         // We own calling init() as it can block.
         if (!reader->init()) {
-            error_string_ += "Failed to init kernel switch reader";
+            error_string_ += "Failed to init kernel " + sequence_type + " reader";
             return sched_type_t::STATUS_ERROR_INVALID_PARAMETER;
         }
     }
     // We assume these sequences are small and we can easily read them all into
     // memory and don't need to stream them on every use.
     // We read a single stream, even if underneath these are split into subfiles
     // in an archive.
-    switch_type_t switch_type = sched_type_t::SWITCH_INVALID;
+    SequenceKey sequence_key;
+    bool in_sequence = false;
     while (*reader != *reader_end) {
         RecordType record = **reader;
         // Only remember the records between the markers.
         trace_marker_type_t marker_type = TRACE_MARKER_TYPE_RESERVED_END;
         uintptr_t marker_value = 0;
-        if (record_type_is_marker(record, marker_type, marker_value) &&
-            marker_type == TRACE_MARKER_TYPE_CONTEXT_SWITCH_START) {
-            switch_type = static_cast<switch_type_t>(marker_value);
-            if (!switch_sequence_[switch_type].empty()) {
-                error_string_ += "Duplicate context switch sequence type found";
+        bool is_marker = record_type_is_marker(record, marker_type, marker_value);
+        if (is_marker && marker_type == start_marker) {
+            sequence_key = static_cast<SequenceKey>(marker_value);
+            in_sequence = true;
+            if (!sequence[sequence_key].empty()) {
+                error_string_ += "Duplicate " + sequence_type + " sequence found";
                 return sched_type_t::STATUS_ERROR_INVALID_PARAMETER;
             }
         }
-        if (switch_type != sched_type_t::SWITCH_INVALID)
```

Comment:
Was this check lost?

Reply:
`switch_type != sched_type_t::SWITCH_INVALID` is same as `in_sequence` now, as far as checking for a previously seen marker_start is concerned. But I guess you're pointing out what if the given template file itself had an incorrect switch_type in some template? I guess that could happen and it would be nice to check for it.

Reply:
There's also the case of a marker_end without a marker_start, and marker_start without the prior seq ending.

I wasn't sure before how much of this invariant checking should be redone here (since the invariant checker already has logic to disallow nested syscall/switch seq; though it doesn't have the wrong-seq-key-in-template-file check). And adding some of that logic here required the template specialized default_seq_key() (couldn't get around the Windows uninit-use warning noted in the comment below), which is slightly tedious because of the prohibition of partial specialization.

But maybe a good idea to do it as the template files don't automatically go through the invariant checker separately.

Reply:
Done.